### DISCUSSION TITLE
CPX usage help

### DISCUSSION THREAD

**Discussion by jmiguelsangar** [bitcraze/discussions#140] [Q&A]
Hi everyone, 
I´ve got a python code which implements some patterns using cflib. I send the code to the crazyflie using crazyradio and it works.
What I want is to send some orders to GAP8 to use the camera during the flight.
Reviewing the documentation about communication protocols I understand that I can use CPX to communicate between STM32, ESP32 and GAP8 in a two way communication.

I have the following questions:

1- The queries sent from the Crazyflie to the GAP8 need to be programed and flash in the STM32? But, how do I connect my python code, is it possible or should I transform python to C++ in order to use CPX within the same code?
2- Should I think to do it the other way round, send requests from GAP8 to STM32 to make it easier?

I´m trying to define the whole process at high level before keep on going...  

Thanks for the help

**Comment by knmcguire** [comment]
Hi! I will try to help but unfortunately those that worked with CPX are not around to help. So I'll need to be quite high level too! The CPX framework does indeed what you describe it in terms of it's plumbing, but there are some elements that are still to be implemented.

1- This needs to be done with the App layer on the STM https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/app_layer/ and then use CPX to send the commands to GAP8. But I'm not sure yet if this framework has already been laid so some coding is probably required.
2- Not sure if this is better though? In depends on your usecase, and if you have a neural network running on the aideck that outputs things that the crazyflie would know about... then yes, gap8 to stm is a good thing to do.

For the latter be aware that CPX is not threadsafe so we are currently having issues with the cluster and cpx, at least for the downlink: https://github.com/bitcraze/aideck-gap8-examples/issues/100
